draft-specifications/Notation.md

```diff
@@ -0,0 +1,218 @@
+- _concrete-string_: _length_ `"` _bytes_
+- _length_: _integer-digits_ :: The number of bytes in _bytes_ as ASCII decimal
+  digits.
+- _bytes_: _byte_* :: The bytes of the string in UTF-8 encoding.
+
+We do not attempt to capture strings with embedded quotes or non-ASCII Unicode
+characters in the abstract notation, but all Unicode strings excluding
+surrogates (U+D800-U+DFFF) are expressible in the concrete representation.
+
+## Selector
+
+A sequence of Unicode code points excluding surrogates (U+D800-U+DFFF).
+
+> Example: `'fleur-de-lis` corresponds to `12'fleur-de-lis`.
+
+- _abstract-selector_: `'` _name_
+- _name_: _alpha_ ( _alpha_ / _digit_ / `-` )*
+- _alpha_: ( `a` - `z` ) / ( `A` - `Z` )
+- _digit_: `0` - `9`
+- _concrete-selector_: _length_ `'` _bytes_
+- _bytes_: _byte_* :: The bytes of the selector in UTF-8 encoding.
+
+We do not attempt to capture selectors with arbitrary Unicode characters in the
+abstract notation, but all Unicode strings excluding surrogates (U+D800-U+DFFF)
+are expressible in the concrete representation.
+
+## ByteArray
+
+An array of 8-bit bytes.
+
+> Example: `:b0b5c0ffeefacade` corresponds to the bytes of `8`, `:`, b0, b5,
+> c0, ff, ee, fa, ca, de in ASCII and hexadecimal.
+
+- _abstract-byte-array_: `:` _hex_
+- _hex_: ( _hex-digit_ _hex-digit_ )*
+- _hex-digit_: _digit_ / ( `a` - `f` ) :: Corresponding to a _byte_ of _bytes_.
+- _concrete-byte-array_: _length_ `:` _bytes_ :: The number of bytes in _bytes_
+  and the _bytes_.
+
+We do not attempt to capture byte arrays with non-space ASCII characters in the
+abstract notation, but all byte arrays are expressible in the concrete
+representation.
+
+## Struct
+
+> The name "struct" is tentative.
+> https://github.com/ocapn/ocapn/pull/125
+
+A collection of unordered (key, value) pairs.
+
+> Examples:
+> - `{ a: 10, b: 2 }` corresponds to `{ 1"a 10+ 1"b 2+ }`.
+> - `{ "a": 10, "b": 2 }` corresponds to `{ 1"a 10+ 1"b 2+ }`.
+> - `{ 'a: 10, 'b: 2 }` corresponds to `{ 1'a 10+ 1'b 2+ }`.
+
+- _abstract-struct_: `{` ( _abstract-field_ ( `,` _abstract-field_ )* )? `}`
+- _abstract-field_: _abstract-key_ `:` _abstract-value_
+- _abstract-key_: _abstract-field-name_ / _abstract-value_
+- _abstract-field-name_: _name_ :: Corresponding to a string.
+- _concrete-struct_: `{` _concrete-field_* `}`
+- _concrete-field_: _concrete-key_ _concrete-value_
+- _concrete-key_: _concrete-value_ :: The concrete representation lacks an analog for the abstraction notation’s string key shorthand.
+
+> The [Model](Model.md) limits field names in structs to strings, but for
+> purposes of [CapTP](CapTP%20Specification.md) surrounding data, the notation
+> and representation allow any value.
+
+The abstract notation allows a shorthand where a field name may be an alphanumeric ASCII
+name without a prefix `"`, in which case the field name is a [String](#string).
+
+## List
+
+A list of any quantity of values.
+
+> Example: `[ 1 2 3 ]` corresponds to `[ 1+ 2+ 3+ ]`.
```

Comment:
The abstract-list suggests items should be comma separated but the example includes no comments. Not sure which is correct, I'm okay either way.

Reply:
I’m fixing this by removing the commas in the example, which makes it consistent with the drafts as written.